File: go/go-fiber-postgres/internal/services/article.go
package services

import (
	"context"
	"database/sql"
	"errors"
	"regexp"
	"strings"
	"time"

	"go.opentelemetry.io/otel/codes"

	"go-fiber-postgres/internal/logging"
	"go-fiber-postgres/internal/models"
	"go-fiber-postgres/internal/repository"
	"go-fiber-postgres/internal/telemetry"
)

var (
	ErrArticleNotFound  = errors.New("article not found")
	ErrNotAuthor        = errors.New("not the author of this article")
	ErrAlreadyFavorited = errors.New("article already favorited")
	ErrNotFavorited     = errors.New("article not favorited")
)

type ArticleService struct {
	articleRepo  *repository.ArticleRepository
	favoriteRepo *repository.FavoriteRepository
}

func NewArticleService(articleRepo *repository.ArticleRepository, favoriteRepo *repository.FavoriteRepository) *ArticleService {
	return &ArticleService{
		articleRepo:  articleRepo,
		favoriteRepo: favoriteRepo,
	}
}

type CreateArticleInput struct {
	Title       string `json:"title"`
	Description string `json:"description"`
	Body        string `json:"body"`
}

type UpdateArticleInput struct {
	Title       *string `json:"title,omitempty"`
	Description *string `json:"description,omitempty"`
	Body        *string `json:"body,omitempty"`
}

type ArticleListResult struct {
	Articles   []*models.Article `json:"articles"`
	TotalCount int               `json:"total_count"`
}

func (s *ArticleService) Create(ctx context.Context, authorID int, input CreateArticleInput) (*models.Article, error) {
	ctx, span := telemetry.Tracer().Start(ctx, "article.create")
	defer span.End()

	slug := generateSlug(input.Title)

	exists, err := s.articleRepo.ExistsBySlug(ctx, slug)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to check slug")
		logging.Error(ctx, "failed to check slug", "error", err)
		return nil, err
	}
	if exists {
		slug = slug + "-" + time.Now().Format("20060102150405")
	}

	article := &models.Article{
		Slug:        slug,
		Title:       input.Title,
		Description: input.Description,
		Body:        input.Body,
		AuthorID:    authorID,
	}

	if err := s.articleRepo.Create(ctx, article); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to create article")
		logging.Error(ctx, "failed to create article", "error", err)
		return nil, err
	}

	telemetry.ArticlesCreated.Add(ctx, 1)
	span.SetStatus(codes.Ok, "article created")
	logging.Info(ctx, "article created", "articleId", article.ID, "slug", slug)

	return s.articleRepo.FindByID(ctx, article.ID)
}

func (s *ArticleService) GetBySlug(ctx context.Context, slug string, userID *int) (*models.Article, error) {
	article, err := s.articleRepo.FindBySlug(ctx, slug)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrArticleNotFound
		}
		return nil, err
	}

	if userID != nil {
		favorited, err := s.favoriteRepo.Exists(ctx, *userID, article.ID)
		if err == nil {
			article.Favorited = favorited
		}
	}

	return article, nil
}

func (s *ArticleService) List(ctx context.Context, limit, offset int, userID *int) (*ArticleListResult, error) {
	articles, err := s.articleRepo.List(ctx, limit, offset)
	if err != nil {
		return nil, err
	}

	count, err := s.articleRepo.Count(ctx)
	if err != nil {
		return nil, err
	}

	if userID != nil {
		favoriteIDs, err := s.favoriteRepo.FindByUserID(ctx, *userID)
		if err == nil {
			favoriteSet := make(map[int]bool)
			for _, id := range favoriteIDs {
				favoriteSet[id] = true
			}
			for _, article := range articles {
				article.Favorited = favoriteSet[article.ID]
			}
		}
	}

	return &ArticleListResult{
		Articles:   articles,
		TotalCount: count,
	}, nil
}

func (s *ArticleService) Update(ctx context.Context, slug string, userID int, input UpdateArticleInput) (*models.Article, error) {
	ctx, span := telemetry.Tracer().Start(ctx, "article.update")
	defer span.End()

	article, err := s.articleRepo.FindBySlug(ctx, slug)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			span.RecordError(ErrArticleNotFound)
			span.SetStatus(codes.Error, ErrArticleNotFound.Error())
			return nil, ErrArticleNotFound
		}
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to find article")
		return nil, err
	}

	if article.AuthorID != userID {
		span.RecordError(ErrNotAuthor)
		span.SetStatus(codes.Error, ErrNotAuthor.Error())
		return nil, ErrNotAuthor
	}

	if input.Title != nil {
		article.Title = *input.Title
		article.Slug = generateSlug(*input.Title)
	}
	if input.Description != nil {
		article.Description = *input.Description
	}
	if input.Body != nil {
		article.Body = *input.Body
	}

	if err := s.articleRepo.Update(ctx, article); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to update article")
		logging.Error(ctx, "failed to update article", "error", err)
		return nil, err
	}

	span.SetStatus(codes.Ok, "article updated")
	logging.Info(ctx, "article updated", "articleId", article.ID)

	return s.articleRepo.FindByID(ctx, article.ID)
}

func (s *ArticleService) Delete(ctx context.Context, slug string, userID int) error {
	ctx, span := telemetry.Tracer().Start(ctx, "article.delete")
	defer span.End()

	article, err := s.articleRepo.FindBySlug(ctx, slug)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			span.RecordError(ErrArticleNotFound)
			span.SetStatus(codes.Error, ErrArticleNotFound.Error())
			return ErrArticleNotFound
		}
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to find article")
		return err
	}

	if article.AuthorID != userID {
		span.RecordError(ErrNotAuthor)
		span.SetStatus(codes.Error, ErrNotAuthor.Error())
		return ErrNotAuthor
	}

	if err := s.articleRepo.Delete(ctx, article.ID); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to delete article")
		logging.Error(ctx, "failed to delete article", "error", err)
		return err
	}

	telemetry.ArticlesDeleted.Add(ctx, 1)
	span.SetStatus(codes.Ok, "article deleted")
	logging.Info(ctx, "article deleted", "articleId", article.ID)

	return nil
}

func (s *ArticleService) Favorite(ctx context.Context, slug string, userID int) (*models.Article, error) {
	ctx, span := telemetry.Tracer().Start(ctx, "article.favorite")
	defer span.End()

	article, err := s.articleRepo.FindBySlug(ctx, slug)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			span.RecordError(ErrArticleNotFound)
			span.SetStatus(codes.Error, ErrArticleNotFound.Error())
			return nil, ErrArticleNotFound
		}
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to find article")
		return nil, err
	}

	exists, err := s.favoriteRepo.Exists(ctx, userID, article.ID)
	if err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to check favorite")
		return nil, err
	}
	if exists {
		span.RecordError(ErrAlreadyFavorited)
		span.SetStatus(codes.Error, ErrAlreadyFavorited.Error())
		return nil, ErrAlreadyFavorited
	}

	favorite := &models.Favorite{
		UserID:    userID,
		ArticleID: article.ID,
	}

	if err := s.favoriteRepo.Create(ctx, favorite); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to create favorite")
		logging.Error(ctx, "failed to create favorite", "error", err)
		return nil, err
	}

	if err := s.articleRepo.IncrementFavorites(ctx, article.ID); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to increment favorites")
		logging.Error(ctx, "failed to increment favorites", "error", err)
		return nil, err
	}

	telemetry.FavoritesAdded.Add(ctx, 1)
	span.SetStatus(codes.Ok, "article favorited")
	logging.Info(ctx, "article favorited", "articleId", article.ID, "userId", userID)

	return s.articleRepo.FindByID(ctx, article.ID)
}

func (s *ArticleService) Unfavorite(ctx context.Context, slug string, userID int) (*models.Article, error) {
	ctx, span := telemetry.Tracer().Start(ctx, "article.unfavorite")
	defer span.End()

	article, err := s.articleRepo.FindBySlug(ctx, slug)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			span.RecordError(ErrArticleNotFound)
			span.SetStatus(codes.Error, ErrArticleNotFound.Error())
			return nil, ErrArticleNotFound
		}
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to find article")
		return nil, err
	}

	if err := s.favoriteRepo.Delete(ctx, userID, article.ID); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			span.RecordError(ErrNotFavorited)
			span.SetStatus(codes.Error, ErrNotFavorited.Error())
			return nil, ErrNotFavorited
		}
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to delete favorite")
		logging.Error(ctx, "failed to delete favorite", "error", err)
		return nil, err
	}

	if err := s.articleRepo.DecrementFavorites(ctx, article.ID); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, "failed to decrement favorites")
		logging.Error(ctx, "failed to decrement favorites", "error", err)
		return nil, err
	}

	telemetry.FavoritesRemoved.Add(ctx, 1)
	span.SetStatus(codes.Ok, "article unfavorited")
	logging.Info(ctx, "article unfavorited", "articleId", article.ID, "userId", userID)

	return s.articleRepo.FindByID(ctx, article.ID)
}

func generateSlug(title string) string {
	slug := strings.ToLower(title)
	reg := regexp.MustCompile(`[^a-z0-9]+`)
	slug = reg.ReplaceAllString(slug, "-")
	slug = strings.Trim(slug, "-")
	return slug
}

File: rust/actix-postgres/src/services/article.rs
use tracing::instrument;

use crate::{
    error::{AppError, AppResult},
    jobs::JobQueue,
    models::{
        ArticleDto, ArticleResponse, ArticlesResponse, CreateArticleInput, ListArticlesQuery,
        UpdateArticleInput,
    },
    repository::{ArticleRepository, FavoriteRepository},
    telemetry::{
        ARTICLES_CREATED, ARTICLES_DELETED, ARTICLES_UPDATED, FAVORITES_ADDED, FAVORITES_REMOVED,
    },
};

#[derive(Clone)]
pub struct ArticleService {
    article_repo: ArticleRepository,
    favorite_repo: FavoriteRepository,
    job_queue: JobQueue,
}

impl ArticleService {
    pub fn new(
        article_repo: ArticleRepository,
        favorite_repo: FavoriteRepository,
        job_queue: JobQueue,
    ) -> Self {
        Self {
            article_repo,
            favorite_repo,
            job_queue,
        }
    }

    #[instrument(name = "article.create", skip(self, input), fields(author_id))]
    pub async fn create(
        &self,
        author_id: i32,
        input: CreateArticleInput,
    ) -> AppResult<ArticleResponse> {
        let slug = self.generate_slug(&input.title);
        let final_slug = if self.article_repo.exists_by_slug(&slug).await? {
            format!(
                "{}-{}",
                slug,
                time::OffsetDateTime::now_utc().unix_timestamp()
            )
        } else {
            slug
        };

        let article = self
            .article_repo
            .create(
                &final_slug,
                &input.title,
                input.description.as_deref().unwrap_or(""),
                &input.body,
                author_id,
            )
            .await?;

        let article_with_author =
            self.article_repo
                .find_by_id(article.id)
                .await?
                .ok_or(AppError::Internal(
                    "Failed to fetch created article".to_string(),
                ))?;

        if let Err(e) = self
            .job_queue
            .enqueue_notification(article.id, &article.title)
            .await
        {
            tracing::warn!(article_id = article.id, error = %e, "Failed to enqueue notification");
        }

        ARTICLES_CREATED.add(1, &[]);

        tracing::info!(article_id = article.id, slug = %article.slug, "Article created");

        Ok(ArticleResponse {
            article: ArticleDto::from_article_with_author(article_with_author, false),
        })
    }

    #[instrument(name = "article.get", skip(self))]
    pub async fn get(&self, slug: &str, user_id: Option<i32>) -> AppResult<ArticleResponse> {
        let article = self
            .article_repo
            .find_by_slug(slug)
            .await?
            .ok_or(AppError::NotFound("Article not found".to_string()))?;

        let favorited = if let Some(uid) = user_id {
            self.favorite_repo.exists(uid, article.id).await?
        } else {
            false
        };

        Ok(ArticleResponse {
            article: ArticleDto::from_article_with_author(article, favorited),
        })
    }

    #[instrument(name = "article.list", skip(self))]
    pub async fn list(
        &self,
        query: ListArticlesQuery,
        user_id: Option<i32>,
    ) -> AppResult<ArticlesResponse> {
        let articles = self
            .article_repo
            .list(query.limit, query.offset, query.author.as_deref())
            .await?;

        let total = self.article_repo.count(query.author.as_deref()).await?;

        let article_ids: Vec<i32> = articles.iter().map(|a| a.id).collect();

        let favorited_ids = if let Some(uid) = user_id {
            self.favorite_repo
                .is_favorited_batch(uid, &article_ids)
                .await?
        } else {
            vec![]
        };

        let articles_dto: Vec<ArticleDto> = articles
            .into_iter()
            .map(|a| {
                let favorited = favorited_ids.contains(&a.id);
                ArticleDto::from_article_with_author(a, favorited)
            })
            .collect();

        Ok(ArticlesResponse {
            articles: articles_dto,
            total,
        })
    }

    #[instrument(name = "article.update", skip(self, input))]
    pub async fn update(
        &self,
        slug: &str,
        user_id: i32,
        input: UpdateArticleInput,
    ) -> AppResult<ArticleResponse> {
        let article = self
            .article_repo
            .find_by_slug(slug)
            .await?
            .ok_or(AppError::NotFound("Article not found".to_string()))?;

        if article.author_id != user_id {
            return Err(AppError::Forbidden);
        }

        let new_slug = input.title.as_ref().map(|t| self.generate_slug(t));

        self.article_repo
            .update(
                article.id,
                new_slug.as_deref(),
                input.title.as_deref(),
                input.description.as_deref(),
                input.body.as_deref(),
            )
            .await?;

        let updated_article =
            self.article_repo
                .find_by_id(article.id)
                .await?
                .ok_or(AppError::Internal(
                    "Failed to fetch updated article".to_string(),
                ))?;

        let favorited = self.favorite_repo.exists(user_id, article.id).await?;

        ARTICLES_UPDATED.add(1, &[]);

        tracing::info!(article_id = article.id, "Article updated");

        Ok(ArticleResponse {
            article: ArticleDto::from_article_with_author(updated_article, favorited),
        })
    }

    #[instrument(name = "article.delete", skip(self))]
    pub async fn delete(&self, slug: &str, user_id: i32) -> AppResult<()> {
        let article = self
            .article_repo
            .find_by_slug(slug)
            .await?
            .ok_or(AppError::NotFound("Article not found".to_string()))?;

        if article.author_id != user_id {
            return Err(AppError::Forbidden);
        }

        self.article_repo.delete(article.id).await?;

        ARTICLES_DELETED.add(1, &[]);

        tracing::info!(article_id = article.id, "Article deleted");

        Ok(())
    }

    #[instrument(name = "article.favorite", skip(self))]
    pub async fn favorite(&self, slug: &str, user_id: i32) -> AppResult<ArticleResponse> {
        let article = self
            .article_repo
            .find_by_slug(slug)
            .await?
            .ok_or(AppError::NotFound("Article not found".to_string()))?;

        let already_favorited = self.favorite_repo.exists(user_id, article.id).await?;

        if !already_favorited {
            self.favorite_repo.create(user_id, article.id).await?;
            self.article_repo.increment_favorites(article.id).await?;
            FAVORITES_ADDED.add(1, &[]);
            tracing::info!(article_id = article.id, user_id, "Article favorited");
        }

        let updated_article = self
            .article_repo
            .find_by_id(article.id)
            .await?
            .ok_or(AppError::Internal("Failed to fetch article".to_string()))?;

        Ok(ArticleResponse {
            article: ArticleDto::from_article_with_author(updated_article, true),
        })
    }

    #[instrument(name = "article.unfavorite", skip(self))]
    pub async fn unfavorite(&self, slug: &str, user_id: i32) -> AppResult<ArticleResponse> {
        let article = self
            .article_repo
            .find_by_slug(slug)
            .await?
            .ok_or(AppError::NotFound("Article not found".to_string()))?;

        let was_favorited = self.favorite_repo.delete(user_id, article.id).await?;

        if was_favorited {
            self.article_repo.decrement_favorites(article.id).await?;
            FAVORITES_REMOVED.add(1, &[]);
            tracing::info!(article_id = article.id, user_id, "Article unfavorited");
        }

        let updated_article = self
            .article_repo
            .find_by_id(article.id)
            .await?
            .ok_or(AppError::Internal("Failed to fetch article".to_string()))?;

        Ok(ArticleResponse {
            article: ArticleDto::from_article_with_author(updated_article, false),
        })
    }

    fn generate_slug(&self, title: &str) -> String {
        generate_slug(title)
    }
}

pub fn generate_slug(title: &str) -> String {
    title
        .to_lowercase()
        .chars()
        .map(|c| if c.is_alphanumeric() { c } else { '-' })
        .collect::<String>()
        .split('-')
        .filter(|s| !s.is_empty())
        .collect::<Vec<_>>()
        .join("-")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_slug_simple() {
        assert_eq!(generate_slug("Hello World"), "hello-world");
    }

    #[test]
    fn test_generate_slug_with_special_chars() {
        assert_eq!(
            generate_slug("Hello, World! How are you?"),
            "hello-world-how-are-you"
        );
    }

    #[test]
    fn test_generate_slug_with_numbers() {
        assert_eq!(
            generate_slug("Top 10 Tips for 2024"),
            "top-10-tips-for-2024"
        );
    }

    #[test]
    fn test_generate_slug_with_multiple_spaces() {
        assert_eq!(
            generate_slug("Multiple   Spaces   Here"),
            "multiple-spaces-here"
        );
    }

    #[test]
    fn test_generate_slug_preserves_lowercase() {
        assert_eq!(generate_slug("UPPERCASE TITLE"), "uppercase-title");
    }

    #[test]
    fn test_generate_slug_handles_leading_trailing_special_chars() {
        assert_eq!(generate_slug("---Hello World---"), "hello-world");
    }

    #[test]
    fn test_generate_slug_empty_title() {
        assert_eq!(generate_slug(""), "");
    }

    #[test]
    fn test_generate_slug_only_special_chars() {
        assert_eq!(generate_slug("!@#$%^&*()"), "");
    }
}
